packages/eslint-plugin/src/rules/no-extraneous-return-types.ts

```diff
@@ -0,0 +1,311 @@
+              return type;
+            })
+            .flatMap(type => tsutils.unionTypeParts(type));
+
+          checkReturnType(arrayTypeArgument, returnTypes);
+        }
+        return;
+      }
+
+      if (node.type === AST_NODE_TYPES.TSUnionType) {
+        for (const typeNode of node.types) {
+          checkReturnType(
+            typeNode,
+            returnTypes,
+            allowUnusedVoidOrUndefinedValues,
+          );
+        }
+
+        return;
+      }
+
+      const declaredReturnType = services.getTypeAtLocation(node);
+
+      const constrainedReturnType = tsutils.isTypeParameter(declaredReturnType)
+        ? checker.getBaseConstraintOfType(declaredReturnType)
+        : declaredReturnType;
+
+      // treat unconstrained type parameters as `unknown` (matches everything)
+      if (!constrainedReturnType) {
+        return;
+      }
+
+      if (
+        allowUnusedVoidOrUndefinedValues &&
+        // don't report on a potentially unused `void` or `undefined` return type
+        (tsutils.isIntrinsicVoidType(constrainedReturnType) ||
+          tsutils.isIntrinsicUndefinedType(constrainedReturnType))
+      ) {
+        return;
+      }
+
+      const hasMatchingReturnStatement = returnTypes.some(actualReturnType => {
+        return checker.isTypeAssignableTo(
+          actualReturnType,
+          constrainedReturnType,
+        );
+      });
+
+      if (!hasMatchingReturnStatement) {
+        context.report({
+          node,
+          messageId: 'unusedReturnTypes',
+          data: {
+            type: checker.typeToString(constrainedReturnType),
+          },
+        });
+      }
+
+      return;
+    }
+
+    function getPotentialReturnTypes(
+      node:
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | TSESTree.FunctionExpression,
+    ): ts.Type[] {
+      if (node.body.type !== AST_NODE_TYPES.BlockStatement) {
+        return tsutils.typeParts(services.getTypeAtLocation(node.body));
+      }
+
+      const nodeTs = services.esTreeNodeToTSNodeMap.get(node);
+
+      const returnTypes: ts.Type[] = [];
+
+      forEachReturnStatement(nodeTs.body as ts.Block, statement => {
+        const expression = statement.expression;
+
+        // `return;`
+        if (!expression) {
+          returnTypes.push(checker.getUndefinedType());
+          return;
+        }
+
+        const typeParts = tsutils.typeParts(
+          checker.getTypeAtLocation(expression),
+        );
+
+        returnTypes.push(...typeParts);
+      });
+
+      return returnTypes;
+    }
+
+    return {
+      'FunctionDeclaration,FunctionExpression,ArrowFunctionExpression'(
+        node:
+          | TSESTree.ArrowFunctionExpression
+          | TSESTree.FunctionDeclaration
+          | TSESTree.FunctionExpression,
+      ): void {
+        if (!node.returnType || node.generator) {
+          return;
+        }
+
+        const actualReturnTypes = getPotentialReturnTypes(node);
```

Comment:
[Style] This confused me at first 😄. Is there a reason to name one _"actual"_ and the other _"potential"_?

If not, request: use the same term for both? 

After reading through it, _"actual"_ matches how I'm understanding it. So +1 from me on that one.